zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/X509ZNodeGroupAclProvider.java

```diff
@@ -0,0 +1,156 @@
+    ZKConfig config = new ZKConfig();
+    this.keyManager = X509AuthenticationUtil.createKeyManager(config);
+    this.trustManager = X509AuthenticationUtil.createTrustManager(config);
+  }
+
+  public X509ZNodeGroupAclProvider(X509TrustManager trustManager, X509KeyManager keyManager) {
+    this.trustManager = trustManager;
+    this.keyManager = keyManager;
+  }
+
+  @Override
+  public KeeperException.Code handleAuthentication(ServerObjs serverObjs, byte[] authData) {
+    ServerCnxn cnxn = serverObjs.getCnxn();
+    X509Certificate clientCert;
+    try {
+      clientCert = X509AuthenticationUtil.getAndAuthenticateClientCert(cnxn, trustManager);
+    } catch (KeeperException.AuthFailedException e) {
+      return KeeperException.Code.AUTHFAILED;
+    }
+
+    // Extract URI from certificate
+    String uri;
+    try {
+      uri = X509AuthenticationUtil.getClientId(clientCert);
+    } catch (Exception e) {
+      // Failed to extract URI from certificate
+      LOG.error(logStrPrefix + "Failed to extract URI from certificate for session 0x{}",
+          Long.toHexString(cnxn.getSessionId()), e);
+      return KeeperException.Code.OK;
+    }
+
+    // User belongs to super user group
+    if (uri.equals(System.getProperty(ZOOKEEPER_ZNODEGROUPACL_SUPERUSER))) {
+      cnxn.addAuthInfo(new Id("super", uri));
+      LOG.info("Authenticated Id '{}' as super user", uri);
+      return KeeperException.Code.OK;
+    }
+
+    // Get authorized domain names for client
+    ClientUriDomainMappingHelper uriDomainMappingHelper =
+        new ZkClientUriDomainMappingHelper(serverObjs.getZks());
+    Set<String> domains = uriDomainMappingHelper.getDomains(uri);
+    if (domains.isEmpty()) {
+      // If no domain name is found, use URI as domain name
+      domains = new HashSet<>();
+      domains.add(uri);
+    }
+
+    Set<String> superUserDomainNames = ZNodeGroupAclProperties.getInstance().getSuperUserDomainNames();
+    for (String domain : domains) {
+      // Grant cross domain components super user privilege
+      if (superUserDomainNames.contains(domain)) {
+        cnxn.addAuthInfo(new Id("super", uri));
+        LOG.info(logStrPrefix + "Id '{}' belongs to superUser domain '{}', authenticated as super user", uri,
+            domain);
+      } else {
+        cnxn.addAuthInfo(new Id(getScheme(), domain));
+        LOG.info(logStrPrefix + "Authenticated Id '{}' for Scheme '{}', Domain '{}'.", uri,
+            getScheme(), domain);
+      }
+    }
+
+    return KeeperException.Code.OK;
+  }
+
+  @Override
+  public boolean matches(ServerObjs serverObjs, MatchValues matchValues) {
+    for (Id id : serverObjs.getCnxn().getAuthInfo()) {
+      if (id.getId().equals(matchValues.getAclExpr()) || id.getId()
+          .equals(System.getProperty(ZOOKEEPER_ZNODEGROUPACL_SUPERUSER))) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  @Override
+  public String getScheme() {
+    return "x509";
+  }
```

Comment:
Question: is using the same scheme as X509AuthProvider okay? How do we make sure X509ZNodeGroupAclProvider is used by the server?

Reply:
It's ok, it's based on x509 certificate that's why the scheme is "x509". For the same reason, we are setting the znode ACL scheme as "x509". When we check zk server log we should be able to tell which authentication provider it is used.